Enable timeout propagation for remote actions

quic-sbjorkle opened the pull request:
When building using remote execution,
action timeouts sometimes need to be adjusted
per action to ensure reasonable timeouts are set.

This change will allow better control over how
actions behave and encourage users to
optimize their actions since the global default timeout does not have to be set to the absolute largest
action timeout in the graph.

sgowroji commented:
Hi @quic-sbjorkle,  Could you please sign the [CLA](https://github.com/bazelbuild/bazel/pull/23961/checks?check_run_id=31440042823).

quic-sbjorkle commented:
> Hi @quic-sbjorkle,  Could you please sign the [CLA](https://github.com/bazelbuild/bazel/pull/23961/checks?check_run_id=31440042823).

Should be signed now if you retrigger the check.

quic-sbjorkle committed:
Allow timeout propagation for remote actions

When building using remote execution,
action timeouts sometimes need to be adjusted
per action to ensure reasonable timeouts are set.

This change will allow better control over how
actions behave and encourage users to
optimize their actions since the global default timeout
does not have to be set to the absolute largest
action timeout in the graph.

Change-Id: I0ea63787ff7769e25dd597575cbfd7fd1050797e

quic-sbjorkle commented:
@tjgq Any ETA for when you can take a look at this? Thanks

brentleyjones commented:
@bazel-io flag

tjgq commented:
I'm supportive of this PR in principle, but I need to do some due diligence to verify that reinterpreting an existing `timeout` tag won't cause problems at Google.

quic-sbjorkle committed:
Add support for timeout:<int> tags to populate ExecutionInfo

Timeout may either be specified as a tag on format
tags = ["tag:<int-seconds>"]
or as execution_requirements on any action on dict format
execution_requirements = {
   "timeout": "<int-seconds>"
}

Change-Id: Ida3bf173a291e7d949046bc4df3d1d4a02d8328d

quic-sbjorkle commented on src/main/java/com/google/devtools/build/lib/packages/TargetUtils.java:
```diff
@@ -58,6 +58,7 @@ private static boolean legalExecInfoKeys(String tag) {
         || tag.startsWith("disable-")
         || tag.startsWith("cpu:")
         || tag.equals(ExecutionRequirements.LOCAL)
+        || tag.equals(ExecutionRequirements.TIMEOUT)
```
It turns out that execution_requirements and tags uses the same map but for tags all values are handled as keys so the value part is discarded.

Wyverald commented:
For triaging purposes -- is this fixing a regression introduced in an earlier 8.0.0 RC?

If not, this need not be release-blocking, but we can certainly cherry-pick to 8.1.0 at least.

tjgq commented:
@Wyverald It's a feature request, no need to hold up the release.

quic-sbjorkle commented:
@tjgq Are there any more changes needed from my side, or is it good to go now?

tjgq commented:
Sorry, I'm about to be out of office, so I won't have time to look into this until January. I would also like @zhengwei143 to do a review of his own.

fmeum commented:
@zhengwei143 @tjgq Friendly ping

tjgq commented on src/main/java/com/google/devtools/build/lib/actions/ExecutionRequirements.java:
```diff
@@ -118,7 +118,31 @@ public String parseIfMatches(String tag) throws ValidationException {
-  public static final String TIMEOUT = "timeout";
+  public static final ParseableRequirement TIMEOUT =
+      ParseableRequirement.create(
+          "timeout:<int>",
```
Can we use a syntax that includes a time unit, e.g. `30s` or `5m` or `1h`, for added clarity? (See `DURATION_REGEX` in src/main/java/com/google/devtools/common/options/Converters.java for inspiration.)

tjgq commented on src/main/java/com/google/devtools/build/lib/packages/TargetUtils.java:
```diff
@@ -58,6 +58,7 @@ private static boolean legalExecInfoKeys(String tag) {
         || tag.startsWith("disable-")
         || tag.startsWith("cpu:")
         || tag.equals(ExecutionRequirements.LOCAL)
+        || tag.startsWith("timeout")
```
This should match either `timeout` exactly, or a `timeout:` prefix (i.e., something like `timeout_blah` shouldn't match).

tjgq commented on src/main/java/com/google/devtools/build/lib/actions/Spawns.java:
```diff
@@ -152,21 +155,72 @@ public static Duration getTimeout(Spawn spawn) throws ExecException {
+                        String.format(
+                            "%s has a '%s' tag, but its value '%s' didn't pass validation: %s",
+                            spawn.getTargetLabel(), requirement.userFriendlyName(), e.getTagValue(), e.getMessage()))
+                    .setSpawn(FailureDetails.Spawn.newBuilder().setCode(Code.INVALID_TIMEOUT_TAG))
```
I think it's fine to reuse `Code.TIMEOUT` here and let the error message carry the additional information.

tjgq commented on src/main/java/com/google/devtools/build/lib/actions/Spawns.java:
```diff
@@ -152,21 +155,72 @@ public static Duration getTimeout(Spawn spawn) throws ExecException {
+        throw new UserExecException(
+            e,
+            FailureDetail.newBuilder()
+                .setMessage("could not parse timeout")
```
Since you're here, can you also include `value` in the error message?

tjgq commented on src/main/java/com/google/devtools/build/lib/actions/Spawns.java:
```diff
@@ -152,21 +155,72 @@ public static Duration getTimeout(Spawn spawn) throws ExecException {
+                throw new UserExecException(
+                    FailureDetail.newBuilder()
+                        .setMessage("multiple timeout values specified")
+                        .setSpawn(FailureDetails.Spawn.newBuilder().setCode(Code.DUPLICATE_TIMEOUT_TAGS))
```
I'm not sure that this is the behavior we want on duplicates. If the rule author says `execution_requirements = {"timeout": "5m"}` but the user says `my_library(tags = ["timeout:10m"])`, wouldn't we want the latter to override the former?

It might be acceptable to always take the more permissive option; otherwise we'd have to track the provenance of execution requirements, which might be hard.

tjgq commented on src/main/java/com/google/devtools/build/lib/exec/StandaloneTestStrategy.java:
```diff
@@ -119,7 +119,7 @@ public TestRunnerSpawn createTestRunnerSpawn(
       executionInfo.put(ExecutionRequirements.NO_CACHE, "");
     }
-    executionInfo.put(ExecutionRequirements.TIMEOUT, "" + getTimeout(action).toSeconds());
+    executionInfo.put("timeout", "" + getTimeout(action).toSeconds());
```
Why change this to a string literal?

Also, it's unclear to me how `tags` interacts with the `timeout` attribute for test targets (or the `--test_timeout` flag, which overrides it). Can you add integration tests to clarify the behavior?

meisterT commented:
What's the status of this PR?

quic-sbjorkle commented:
> What's the status of this PR?

I had some issues understanding the proper places where to place unit-tests so they actually verify the remote execution requirements. And then I haven't had time to pick it up again. I will try to squeeze in some time to finish this up before summer.

quic-sbjorkle committed:
Merge codelinaro/master into allow_remote_action_timeout_propagation

Change-Id: I8c4236c8d68bfc7b171dc24033429525b6484e45

quic-sbjorkle committed:
Adjust implementation according to PR review comments

Change-Id: Idc57b3574f2be888338451c602bc338dcb3b66d6

quic-sbjorkle committed:
Cleanup dead code and adjust indentation

Change-Id: Idacb51f8cc9af7f09598525fb5ff4257dc4c8c0c

quic-sbjorkle commented:
@tjgq Let me know if you have any more comments. Thanks!

quic-sbjorkle commented:
@meisterT friendly ping :-)

quic-sbjorkle commented:
@tjgq @meisterT , This is a blocker for us to move over to using an unpatched bazel so would be greatful to get this into an official release.

iancha1992 commented:
@bazel-io flag

iancha1992 commented:
@bazel-io fork 9.2.0

EdSchouten commented:
How does this PR relate to https://github.com/bazelbuild/bazel/pull/29395 ? It looks like this is adding the timeout option as an execution parameter, while #29395 adds it as a dedicated argument to `ctx.actions.run()`. Which of those approaches is preferred?
